### ISSUE TITLE
raw query does not seem to support the SOQL "__r" notation

### ISSUE THREAD

**Issue by dalem1959** [django-salesforce/django-salesforce#95]
I have some existing SOQL code that is rather complex and was hoping to use in a .raw() call.  It appears that the SOQL "__r" notation is not supported:

I have a field on our Asset object, Responsible_SE__c which is of type Lookup(User), that is, a foreign key into the User object.  The proper SOQL to see the name of the Responsible_SE__c User would be Responsible_SE__r.Name.

But when I run this code

`for asset in Asset.objects.raw(''' Select a.Id, a.Responsible_SE__r.Name FROM Asset a'''):`
`...     asset
`

I get the following error:

> Traceback (most recent call last):
>  File "<console>", line 1, in <module>
>  File "/opt/python2.7/lib/python2.7/site-packages/django/db/models/query.py", line 1512, in **iter**
>     values = compiler.resolve_columns(values, fields)
>  File "/u/dale/ds/salesforce/backend/compiler.py", line 25, in resolve_columns
>    return [row[field.column] for field in fields]
> AttributeError: 'NoneType' object has no attribute 'column'

Similarly, if I try
`Asset.objects.all().values('serialnumber', 'responsible_se__r__name')`

I get the error:

> Traceback (most recent call last):
>  File "<console>", line 1, in <module>
>  File "/opt/python2.7/lib/python2.7/site-packages/django/db/models/query.py", line 578, in values
>     return self._clone(klass=ValuesQuerySet, setup=True, _fields=fields)
>  File "/opt/python2.7/lib/python2.7/site-packages/django/db/models/query.py", line 881, in _clone
>     c._setup_query()
>  File "/opt/python2.7/lib/python2.7/site-packages/django/db/models/query.py", line 1005, in _setup_query
>    self.query.add_fields(self.field_names, True)
>  File "/opt/python2.7/lib/python2.7/site-packages/django/db/models/sql/query.py", line 1669, in >add_fields
>    "Choices are: %s" % (name, ", ".join(names)))
> FieldError: Cannot resolve keyword 'responsible_se__r__name' into field. Choices are: account, account_owner, etc.

**Comment by philchristensen** [comment]
I believe this is less an issue with raw() and move of an issue with SOQL in general. There's more about it in #43

**Comment by hynekcer** [comment]
I think that Django also doesn't support raw queries based on Model.objects with more tables, except annotation (Count, Sum etc.) based on a related table. It is clear that the same field names can clash between tables. It is also not tested by Django, except the case above "Django-1.7.4/tests/raw_query":

    SELECT a.*, count(b.id) as book_count FROM raw_query_author a
        LEFT JOIN raw_query_book b ON a.id = b.author_id
        GROUP BY a.id, a.first_name, a.last_name, a.dob ORDER BY a.id

I wrote an improvement of raw queries in November but I terminated it because I don't want make django-salesforce better than Django (incompatibile).

A current solution for you is a a low level raw query.

    from django.conf import connections

    def my_raw(soql, args):
        cur = connections['salesforce'].cursor()
        cur.execute(soql, args)
        for x in cursor():
            yield x

The cursor is created fast without additional network activity. The fields must be accessed by dict with Salesforce API names used as keys. I think it is better for SOQL raw:

    for asset in Asset.objects.raw(''' Select a.Id, a.Responsible_SE__r.Name FROM Asset a'''):
        print(asset['Id'], asset['Responsible_SE__c']['Name'])

Arguments with `%s` are supported by SF without qoutes and without manual escaping:

     cur.execute("select name from Contact where id = %s", ('003M000000YhJhiIAF',))

I'm open to improvements of this type of raw query. (I can imagine to buld partial model objects with fields that can be accessed by dot and Django field name much easier than for Model.objects.raw queries.)

The suffix '__r' has been solved in the development version for filters ("WHERE" part of the query), but for values ("SELECT" part it is still a problem without a known solution.

By the way, you need not to try '__r' ('responsible_se__r') where a Django name is expected, because everything is splittet to parts at double underscore by Django.

**Comment by hynekcer** [comment]
I was working on this last week, because I also want to close it soon.

**Comment by philchristensen** [comment]
Oops, sorry, I thought this was a user error situation, didn't read it closely enough I guess.

**Comment by hynekcer** [comment]
Fixed:
example 1:
```
    >>> soql = ''' Select Asset.Id, Asset.Responsible_SE__r.Name FROM Asset'''
    >>> qs = Asset.objects.raw(soql)
    >>> list(soql)[0].__dict__.keys()
    [..., 'Responsible_SE__r.Name',...]
    >>> qs = Asset.objects.raw(soql, translations={'Responsible_SE__r.Name':
                                                   'my_field_responsible_name'})
    >>> print(qs[0].my_field_responsible_name)
```
You can frequently omit the table name in field names or use the full table name "Asset", but an alias "a" for the table name can be never used in SOQL.

example 2)  
    `Asset.objects.all().values('serialnumber', 'responsible_se__name')`  
All names in `.values` are Django model names, usually lowercase and never with `'__c'` suffix.

**Comment by hynekcer** [comment]
Fixed by PR #222
